lib/Binding.lua

```diff
@@ -0,0 +1,125 @@
+local Symbol = require(script.Parent.Symbol)
+local Type = require(script.Parent.Type)
+
+local createSignal = require(script.Parent.createSignal)
+
+--[[
+	Markers for fields and methods that are hidden from external users
+]]
+local Internal = {
+	changeSignal = Symbol.named("changeSignal"),
+	value = Symbol.named("value"),
+
+	update = Symbol.named("update"),
+	subscribe = Symbol.named("subscribe"),
+}
+
+local bindingMetatable = {
+	__tostring = function(self)
+		return ("RoactBinding(%s)"):format(tostring(self[Internal.value]))
+	end
+}
+
+local Binding = {}
+
+--[[
+	Create a new binding object with the given starting value. This
+	function will be exposed to users of Roact.
+]]
+function Binding.create(initialValue)
+	local binding = Binding.createFromSource(initialValue, function(value)
+		return value
+	end)
+
+	local updater = function(newValue)
+		binding[Internal.update](binding, newValue)
+	end
+
+	return binding, updater
+end
+
+--[[
+	Creates a new binding from the given source and with the given mapping.
+
+	The source can either be a starting value or another Binding object,
+	which is useful for mapping a binding onto another binding
+]]
+function Binding.createFromSource(source, mapFunc)
+	local initialValue = source
+	if Type.of(source) == Type.Binding then
+		initialValue = source:getValue()
+	end
+
+	local subCount = 0
+	local disconnectSource = nil
+
+	local binding = {
+		[Type] = Type.Binding,
+
+		[Internal.value] = mapFunc(initialValue),
+		[Internal.changeSignal] = createSignal(),
+	}
+
+	binding[Internal.update] = function(self, newValue)

```

Comment:
Yeah, that's probably the way to go. The thing that made this approach convenient is that I could use upvalues instead of making everything private via Symbols, which is an elegant solution conceptually but it makes the bodies of my functions kind of hideous.